### PR TITLE
Fixes #190: Create C tutorial: Getting the default stream for a module 

### LINKED ISSUE #190: Create C tutorial: Getting the default stream for a module
Currently, the README.md and [modulemd.h](https://github.com/fedora-modularity/libmodulemd/blob/master/modulemd/v2/include/modulemd-2.0/modulemd.h) files have some limited guides for how to work with libmodulemd, but some real tutorials on API 2.0 would be ideal.

Create tutorial documentation (either extending modulemd.h annotations or by adding a new Tutorial section to the documentation) describing the following process with examples:

* Include the appropriate C headers
* Read repodata YAML from a file into an index
* Query the `Modulemd.ModuleIndex` for the module with a given name
* Query the `Modulemd.Module` for the default stream.

### PR DESCRIPTION
**Pull request by OrionStar25**
Fixes #190  
I couldn't find:

1.  The `yaml` files of "update" and "fedora" repo.
2.  `get_rpm_api() ` and `get_dependencies()` function for the `stream `structure.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,10 +22,17 @@ from gi.repository import Modulemd
 The libmodulemd API provides a number of convenience tools for interacting
 with repodata (that is, streams of YAML that contains information on multiple
 streams, default data and translations). The documentation will use two
-repositories, called "fedora" and "updates" for demonstrative purposes. It
-will assume that the content of the YAML module metadata from those two
-repositories have been loaded into string variables "fedora_yaml" and
-"updates_yaml", respectively.
+repositories, called "fedora" and "updates" for demonstrative purposes.
```
Comment (sgallagher) [inline README.md]: As in the Python version, please don't replace this section.

File: README.md [hunk 2]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
 upgraded internally).
 
+Now, we can start operating on the retrieved data. This guide will
+give only a brief overview of the most common operations. See the API
+specification for a full list of information that can be retrieved.
+
+## Discover the default stream for a particular module.
+## C
+```C
+ module =  modulemd_module_index_get_module(merged_index,'modulename');
```
Comment (sgallagher) [inline README.md]:
```suggestion
 module = modulemd_module_index_get_module (merged_index, "modulename");
```

File: README.md [hunk 3]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+ defaults = modulemd_module_get_defaults (module);
+ printf ("Default stream for modulename is %s",
```
Comment (sgallagher) [inline README.md]:
```suggestion
 printf ("Default stream for modulename is %s\n", 
```

File: README.md [hunk 4]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+ module =  modulemd_module_index_get_module(merged_index,'modulename');
+ defaults = modulemd_module_get_defaults (module);
+ printf ("Default stream for modulename is %s", 
+       modulemd_defaults_get_module_name (defaults));
+  ```
+ 
+ ## Get the list of RPMs defining the public API for a particular module NSVC
+ ## C
+```C
+module =  modulemd_module_index_get_module(merged_index,'modulename');
```
Comment (sgallagher) [inline README.md]:
```suggestion
module = modulemd_module_index_get_module (merged_index, "modulename");
```

File: README.md [hunk 5]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+stream = modulemd_module_get_stream_by_NSVC(module, 'modulestream', 1, 'deadbeef');
```
Comment (sgallagher) [inline README.md]:
```suggestion
stream = modulemd_module_get_stream_by_NSVC (module, "modulestream", 1, "deadbeef");
```

File: README.md [hunk 6]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+api_list = stream.get_rpm_api() //change this
```
Comment (sgallagher) [inline README.md]: This hasn't been updated from the Python example.

File: README.md [hunk 7]
```diff
@@ -60,7 +67,7 @@ indexes into a combined one:
 
 ## C
 ```C
-merger = modulemd_module_index_merger_new()
+merger = modulemd_module_index_merger_new();
```
Comment (sgallagher) [inline README.md]: General comment that applies to all of the assignments below:

Unlike Python, you need to declare your C variables ahead of time. In this case, you need `ModulemdModuleIndexMerger *merger =`

Do the same for all of the examples.

File: README.md [hunk 8]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+```C
+module =  modulemd_module_index_get_module(merged_index,'modulename');
+stream = modulemd_module_get_stream_by_NSVC(module, 'modulestream', 1, 'deadbeef');
+api_list = stream.get_rpm_api() //change this
+```
+
+ ## Retrieve the modular runtime dependencies for a particular module NSVC
+## C
+```C
+module =  modulemd_module_index_get_module(merged_index,'modulename');
```
Comment (sgallagher) [inline README.md]:
```suggestion
module = modulemd_module_index_get_module(merged_index, "modulename");
```

File: README.md [hunk 9]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+stream = modulemd_module_get_stream_by_NSVC(module, 'modulestream', 1, 'deadbeef');
```
Comment (sgallagher) [inline README.md]:
```suggestion
stream = modulemd_module_get_stream_by_NSVC(module, "modulestream", 1, "deadbeef");
```

File: README.md [hunk 10]
```diff
@@ -87,6 +94,39 @@ example if the repodata has a mix of v1 and v2 ModuleStream objects, the
...
+//change this
```
Comment (sgallagher) [inline README.md]: Do as it says :smile:

File: README.md (hunk 7 above)
Comment (OrionStar25) [reply README.md]: The current `README` doesn't contain a declaration for C variables, so I thought to follow uniformity. Should I change it for them too then? I thought its implied.

File: README.md (hunk 6 above)
Comment (OrionStar25) [reply README.md]: I couldn't find any function in C for this and `stream.get_dependencies()` . Can you point me towards it?

**Comment by OrionStar25** [comment]
@sgallagher  I made the syntax changes

File: README.md (hunk 7 above)
Comment (sgallagher) [reply README.md]: Hmm, you're correct. The existing README probably should be updated. You don't need to do so as part of this PR.

File: README.md (hunk 6 above)
Comment (sgallagher) [reply README.md]: [`modulemd_module_stream_v2_get_rpm_api_as_strv()`](https://sgallagh.fedorapeople.org/docs/libmodulemd/2.0/modulemd-2.0-Modulemd.ModuleStreamV2.html#modulemd-module-stream-v2-get-rpm-api-as-strv) and [`modulemd_module_stream_v2_get_dependencies ()`](https://sgallagh.fedorapeople.org/docs/libmodulemd/2.0/modulemd-2.0-Modulemd.ModuleStreamV2.html#modulemd-module-stream-v2-get-dependencies)

**Comment by OrionStar25** [comment]
@sgallagher `modulemd_module_stream_v2_get_dependencies ()` returns a GPtrPointer, so I'm not sure if getting the size of the struct array is right the way I did it. Or do we store `NULL` in the end so that I could iterate over it to check for it?

**Comment by sgallagher** [comment]
@OrionStar25 A `GPtrArray` has a member attribute `len` that you can use. See https://developer.gnome.org/glib/stable/glib-Pointer-Arrays.html#GPtrArray

File: README.md [hunk 11]
```diff
@@ -109,6 +122,20 @@ specification for a full list of information that can be retrieved.
...
 
  ## Retrieve the modular runtime dependencies for a particular module NSVC
+## C
+```C
+module = modulemd_module_index_get_module (merged_index, "modulename");
+stream = modulemd_module_get_stream_by_NSVC (module, "modulestream", 1, "deadbeef");
+deps_list = modulemd_module_stream_v2_get_dependencies (stream);
+int length = sizeof(deps_list)/sizeof(deps_list[0]);
+
+for(int i = 0; i < length;i++)
```
Comment (sgallagher) [inline README.md]: This isn't how GPtrArray lists work. You will want to do
```
for (gint i = 0; i < deps_list->len; i++)
  {
    stuff with g_ptr_array_index(deps_list, i);
  }
```

**Comment by OrionStar25** [comment]
@sgallagher  done!

File: README.md (hunk 7 above)
Comment (sgallagher) [reply README.md]: Sorry, I was unclear with this message. You don't need to fix the existing README to have the variable declarations, but I would like you to add them for the new tutorial you are writing. We can fix the lack of them from the existing section in a future PR.

**Comment by OrionStar25** [comment]
@sgallagher I declared variables for the existing c tutorial too.

**Comment by OrionStar25** [comment]
FORCE-PUSH FOR THE WIN! Meld was creating 4 different files, i didn't understand what to do with them.

File: README.md [hunk 12]
```diff
@@ -118,7 +145,7 @@ specification for a full list of information that can be retrieved.
     depstream_list = dep.get_runtime_streams('depstreamname')
     <do_stuff>
 ```
-
+
```
Comment (sgallagher) [inline README.md]: Could you delete this extra space and force-push?

Comment (OrionStar25) [reply README.md]: Done

File: README.md [hunk 13]
```diff
@@ -118,7 +145,6 @@ specification for a full list of information that can be retrieved.
     depstream_list = dep.get_runtime_streams('depstreamname')
     <do_stuff>
 ```
-
```
Comment (sgallagher) [inline README.md]: Leave the newline here, please. The previous request was just to eliminate the unexpected whitespace you added.

Other than that, assuming the CI tests all pass, I think this is probably ready to merge.

Comment (OrionStar25) [reply README.md]: Ah, my bad. 2 minutes

**Comment by sgallagher** [comment]
Approved! Merging it now. Thanks for your contribution!

**Pull request merged**